### PR TITLE
Support exception stack traces across RPC boundaries

### PR DESCRIPTION
**Pull request by kentonv**
It is up to the application to register a callback which will be used to encode the stack trace for remote consumption if desired. If no callback is given, no new information is sent. This allows apps to decide what information they are comfortable revealing to the client, as well as to add app-specific information like an instance name or version.

### CODE REVIEW

File: c++/src/kj/exception.c++ [hunk 1]
```diff
@@ -752,6 +756,10 @@ Exception::Exception(const Exception& other) noexcept
     file = ownFile.cStr();
   }
 
+  if (other.remoteTrace != nullptr) {
```
Comment (a-robinson) [inline c++/src/kj/exception.c++:759]: Huh, I hadn't noticed this functionality of comparing a `String` to `nullptr` before.

File: c++/src/capnp/rpc.c++ [hunk 2]
```diff
@@ -112,11 +112,16 @@ Orphan<List<rpc::PromisedAnswer::Op>> fromPipelineOps(
...
+  if (exception.hasTrace()) {
+    result.setRemoteTrace(kj::str(exception.getTrace()));
+  }
+  return result;
 }
 
-void fromException(const kj::Exception& exception, rpc::Exception::Builder builder) {
+void fromException(const kj::Exception& exception, rpc::Exception::Builder builder,
+                   kj::Maybe<kj::Function<kj::String(const kj::Exception&)>&> traceEncoder) {
   // TODO(someday):  Indicate the remote server name as part of the stack trace.  Maybe even
```
Comment (a-robinson) [inline c++/src/capnp/rpc.c++:125]: Looks like you can knock off the "Maybe even transmit stack traces" part of this TODO

File: c++/src/capnp/rpc.c++ [hunk 3]
```diff
@@ -319,6 +329,11 @@ public:
   }
 
   void disconnect(kj::Exception&& exception) {
+    // After disconnect(), the RpcSystem could be destoryed, making `traceEncoder` a dangling
```
Comment (a-robinson) [inline c++/src/capnp/rpc.c++:332]: minor typo on destoryed

**Review by a-robinson — APPROVED** [review]
Very nice, thanks for adding this.

**Comment by kentonv** [comment]
Fixups: https://github.com/capnproto/capnproto/commit/58a4a0f196d6136e128ba4526920f79307c4ed1b

**Pull request merged**